// monitorctrlc/index.js
"use strict";
var chalk = require("chalk");
var os = require("os");
var emptyDisposable = {
    dispose: function () { return; }
};
var constants = os.constants; // not present in node4 and mis-typed in node6
var sigint = (constants && constants.signals && typeof constants.signals.SIGINT === "number"
    ? constants.signals.SIGINT
    : 2);
/**
 * Exit code returned when SIGINT causes termination.
 *
 * As per node docs[0], will be 128 + value of SIGINT (or 2
 * if not available).
 *
 * [0]: https://nodejs.org/dist/latest-v6.x/docs/api/process.html#process_exit_codes
 */
exports.sigintExitCode = 128 + sigint;
/**
 * Prints a message indicating Ctrl+C was pressed then exits the process.
 *
 * Exit code will be value of "sigintExitCode".
 */
function defaultCtrlCHandler() {
    console.log("'" + chalk.cyan("^C") + "', exiting");
    process.exit(exports.sigintExitCode);
}
exports.defaultCtrlCHandler = defaultCtrlCHandler;
/**
 * Monitors Ctrl+C and executes a callback instead of SIGINT.
 *
 * When invoked, this listens to keypresses on STDIN,
 * which prevents normal process termination. To ensure
 * your program terminates properly if Ctrl+C is NOT
 * pressed, call the "dispose" method on the object
 * returned by this function when your process is
 * completed.
 *
 * NOTE: This should only be used by programs that do
 * not normally read from STDIN.
 *
 * @param {Function} onCtrlC
 *     Callback function to execute on Ctrl+C.
 *     @default "defaultCtrlCHandler"
 * @returns
 *     Disposable object that restores STDIN to its previous
 *     state, allowing for proper process termination.
 */
function monitorCtrlC(onCtrlC) {
    if (onCtrlC === void 0) { onCtrlC = defaultCtrlCHandler; }
    function monitorCtrlCOnData(data) {
        if (data.length === 1 && data[0] === 0x03) {
            onCtrlC();
        }
    }
    var stdin = process.stdin;
    if (stdin && stdin.isTTY) {
        stdin.setRawMode(true);
        stdin.on("data", monitorCtrlCOnData);
        return {
            dispose: function () {
                stdin.removeListener("data", monitorCtrlCOnData);
                stdin.setRawMode(false);
                stdin.pause();
            }
        };
    }
    return emptyDisposable;
}
exports.monitorCtrlC = monitorCtrlC;
